x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java

```diff
@@ -0,0 +1,290 @@
+import org.elasticsearch.lucene.spatial.CartesianShapeDocValuesQuery;
+import org.elasticsearch.search.sort.NestedSortBuilder;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+import org.elasticsearch.xpack.ql.QlIllegalArgumentException;
+import org.elasticsearch.xpack.ql.querydsl.query.Query;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.CARTESIAN_POINT;
+
+public class SpatialRelatesQuery extends Query {
+    private final String field;
+    private final ShapeField.QueryRelation queryRelation;
+    private final Geometry shape;
+    private final DataType dataType;
+
+    public SpatialRelatesQuery(Source source, String field, ShapeField.QueryRelation queryRelation, Geometry shape, DataType dataType) {
+        super(source);
+        this.field = field;
+        this.queryRelation = queryRelation;
+        this.shape = shape;
+        this.dataType = dataType;
+    }
+
+    @Override
+    public boolean containsNestedField(String path, String field) {
+        return false;
+    }
+
+    @Override
+    public Query addNestedField(String path, String field, String format, boolean hasDocValues) {
+        return null;
+    }
+
+    @Override
+    public void enrichNestedSort(NestedSortBuilder sort) {
+
+    }
+
+    @Override
+    public QueryBuilder asBuilder() {
+        return EsqlDataTypes.isSpatialGeo(dataType) ? new GeoShapeQueryBuilder() : new CartesianShapeQueryBuilder();
+    }
+
+    @Override
+    protected String innerToString() {
+        throw new IllegalArgumentException("SpatialRelatesQuery.innerToString() not implemented");
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field, queryRelation, shape, dataType);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        SpatialRelatesQuery other = (SpatialRelatesQuery) obj;
+        return Objects.equals(field, other.field)
+            && Objects.equals(queryRelation, other.queryRelation)
+            && Objects.equals(shape, other.shape)
+            && Objects.equals(dataType, other.dataType);
+    }
+
+    public ShapeRelation shapeRelation() {
+        return switch (queryRelation.toString().toLowerCase(Locale.ROOT)) {
+            case "intersects" -> ShapeRelation.INTERSECTS;
+            case "disjoint" -> ShapeRelation.DISJOINT;
+            case "within" -> ShapeRelation.WITHIN;
+            case "contains" -> ShapeRelation.CONTAINS;
+            default -> throw new QlIllegalArgumentException("Unknown shape relation [{}]", queryRelation);
+        };
```

Comment:
It is unnecessary to use strings here, you can do it like that? : 

```
 return switch (queryRelation) {
            case INTERSECTS -> ShapeRelation.INTERSECTS;
            case DISJOINT -> ShapeRelation.DISJOINT;
            case WITHIN -> ShapeRelation.WITHIN;
            case CONTAINS -> ShapeRelation.CONTAINS;
        };
```

Reply:
Good point. I've been ruined by too many cases where the static type is not a proper `enum` like `DataType` and `EsqlDataType`, where we have to use strings. But since these types are proper `enum` types, I'm fixed this like you suggest.